lib/passes/analysis/MemOpVisitor.cpp

```diff
@@ -78,65 +99,85 @@ void MemOpVisitor::visitMallocLike(llvm::CallBase& ci, MemOpKind k) {
             }
           }
         }
       }
     }
-    // FIXME this is a try to fix issue #13 ; may require more sophisticated dataflow tracking
+    // GEP indicates that an array cookie is added to the allocation. (Fixes #13)
     if (auto gep = dyn_cast<GetElementPtrInst>(user)) {
-      // if (gep->getPointerOperand() == ci) {
-      for (auto gep_user : gep->users()) {
-        if (auto bcastInst = dyn_cast<BitCastInst>(gep_user)) {
-          bcasts.insert(bcastInst);
-        }
-      }
-      //}
+      geps.insert(gep);
     }
   }
+  return {geps, bcasts};
+}
 
-  BitCastInst* primary_cast{nullptr};
-  if (!bcasts.empty()) {
-    primary_cast = *bcasts.begin();
-  }
+auto handleUnpaddedArrayCookie(MallocGeps const& geps, MallocBcasts& bcasts, BitCastInst*& primary_cast)
+    -> llvm::Optional<ArrayCookieData> {
+  using namespace util::type;
+  // We expect only the bitcast to size_t for the array cookie store.
+  assert(bcasts.size() == 1);
+  auto cookie_bcast = *bcasts.begin();
+  assert(isi64Ptr(cookie_bcast->getDestTy()));
+  auto cookie_store = expectSingleUser<StoreInst>(cookie_bcast);
+
+  auto array_gep = *geps.begin();
+  assert(array_gep->getNumIndices() == 1);
+  auto array_bcast = expectSingleUser<BitCastInst>(array_gep);
+  bcasts.insert(array_bcast);
+  primary_cast = array_bcast;
+
+  return {ArrayCookieData{cookie_store, array_gep}};
+}
 
+auto handlePaddedArrayCookie(MallocGeps const& geps, MallocBcasts& bcasts, BitCastInst*& primary_cast)
+    -> llvm::Optional<ArrayCookieData> {
   using namespace util::type;
-  // const auto is_i64 = [](auto* type) { return type->isPointerTy() && type->getPointerElementType()->isIntegerTy(64);
-  // };
-
-  const bool has_specific = llvm::count_if(bcasts, [&](auto bcast) {
-                              auto dest = bcast->getDestTy();
-                              return !isVoidPtr(dest) && !isi64Ptr(dest);
-                            }) > 0;
-
-  for (auto bcast : bcasts) {
-    auto dest = bcast->getDestTy();
-    if (!isVoidPtr(dest)) {
-      auto cast = bcast;
-      if (isi64Ptr(dest) && has_specific) {  // LLVM likes to treat mallocs with i64 ptr type, skip if we have sth else
-        continue;
-      }
-      primary_cast = cast;
-    }
+  // We expect bitcasts only after the GEP instructions in this case.
+  assert(bcasts.size() == 0);
+
+  auto gep_it     = geps.begin();
+  auto array_gep  = *gep_it++;
+  auto cookie_gep = *gep_it++;
+
+  auto cookie_bcast = expectSingleUser<BitCastInst>(cookie_gep);
+  assert(isi64Ptr(cookie_bcast->getDestTy()));
+  auto cookie_store = expectSingleUser<StoreInst>(cookie_bcast);
+
+  assert(array_gep->getNumIndices() == 1);
+  auto array_bcast = expectSingleUser<BitCastInst>(array_gep);
+  bcasts.insert(array_bcast);
+  primary_cast = array_bcast;
+
+  return {ArrayCookieData{cookie_store, array_gep}};
+}
+
+auto handleGepInstrs(MallocGeps const& geps, MallocBcasts& bcasts, BitCastInst*& primary_cast)
```

Comment:
- `T const& foo` is not a style used in TypeART. Suggestion: e.g., `const T& foo`
- `T*& foo` seems unusual. Preference: `T*`